File: src/linked/double_to_stack_and_queue.java
package linked;

import java.util.LinkedList;
import java.util.Queue;
import java.util.Stack;

class Solution {

  private static class Node<T> {
    public T value;
    public Node<T> pre;
    public Node<T> next;

    public Node(T value) {
      this.value = value;
    }
  }

  /**
   * 双向链表实现双端队列
   */
  public static class DoubleEndsQueue<T> {
    public Node<T> head, tail;

    /**
     * 插入数据到队列头部
     * 
     * @param value
     */
    public void addFromHead(T value) {
      Node<T> cur = new Node<T>(value);
      if (head == null) {
        head = cur;
        tail = cur;
      } else {
        cur.next = head;
        head.pre = cur;

        head = cur;
      }
    }

    /**
     * 插入数据到队列尾部
     * 
     * @param value
     */
    public void addFromBottom(T value) {
      Node<T> cur = new Node<T>(value);
      if (head == null) {
        head = cur;
        tail = cur;
      } else {
        cur.pre = tail;
        tail.next = cur;

        tail = cur;
      }
    }

    /**
     * 双端队列 对头出队
     * 
     * @return
     */
    public T popFromHead() {
      if (head == null)
        return null;
      Node<T> cur = head;
      // 队列中只有一个节点
      if (head == tail) {
        head = null;
        tail = null;
      } else {
        head = head.next;

        cur.next = null;
        head.pre = null;
      }
      return cur.value;
    }

    /**
     * 双端队列 对尾出队
     * 
     * @return
     */
    public T popFromBottom() {
      if (head == null)
        return null;

      Node<T> cur = head;
      // 队列只有一个元素的情况
      if (head == tail) {
        head = null;
        tail = null;
      } else {
        tail = tail.pre;

        tail.next = null;
        cur.pre = null;
      }
      return cur.value;
    }

    public boolean isEmpty() {
      return head == null;
    }

  }

  public static class MyStack<T> {
    private DoubleEndsQueue<T> queue;

    public MyStack() {
      queue = new DoubleEndsQueue<T>();
    }

    public void push(T value) {
      queue.addFromHead(value);
    }

    public T pop() {
      return queue.popFromHead();
    }

    public boolean isEmpty() {
      return queue.isEmpty();
    }
  }

  public static class MyQueue<T> {
    private DoubleEndsQueue<T> queue;

    public MyQueue() {
      queue = new DoubleEndsQueue<T>();
    }

    public void push(T value) {
      queue.addFromBottom(value);
    }

    public T poll() {
      return queue.popFromHead();
    }

    public boolean isEmpty() {
      return queue.isEmpty();
    }
  }

  // ============ TEST ============
  public boolean isEqual(Integer o1, Integer o2) {
    if (o1 == null && o2 != null) {
      return false;
    }
    if (o1 != null && o2 == null) {
      return false;
    }
    if (o1 == null && o2 == null) {
      return true;
    }
    return o1.equals(o2);
  }

  public static void main(String[] args) {
    Solution solution = new Solution();

    int oneTestDataNum = 100;
    int value = 10000;
    int testTimes = 100000;

    for (int i = 0; i < testTimes; i++) {

      MyStack<Integer> myStack = new MyStack<>();
      MyQueue<Integer> myQueue = new MyQueue<>();
      Stack<Integer> stack = new Stack<>();
      Queue<Integer> queue = new LinkedList<>();
      for (int j = 0; j < oneTestDataNum; j++) {
        int nums = (int) (Math.random() * value);
        if (stack.isEmpty()) {
          myStack.push(nums);
          stack.push(nums);
        } else {
          if (Math.random() < 0.5) {
            myStack.push(nums);
            stack.push(nums);
          } else {
            if (!solution.isEqual(myStack.pop(), stack.pop())) {
              System.out.println("oops!");
            }
          }
        }
        int numq = (int) (Math.random() * value);
        if (queue.isEmpty()) {
          myQueue.push(numq);
          queue.offer(numq);
        } else {
          if (Math.random() < 0.5) {
            myQueue.push(numq);
            queue.offer(numq);
          } else {
            if (!solution.isEqual(myQueue.poll(), queue.poll())) {
              System.out.println("oops!");
            }
          }
        }
      }
    }
    System.out.println("finish!");
  }
}

File: src/linked/double_to_stack_and_queue.cpp
#include <iostream>
#include <queue>
#include <stack>
using namespace std;

struct Node {
  int value;
  Node *next;
  Node *last;
  Node() : value(0), next(nullptr), last(nullptr) {}
  Node(int x) : value(x), next(nullptr), last(nullptr) {}
  Node(int x, Node *next, Node *last) : value(x), next(next), last(last) {}
};

struct DoubleEndsQueue {
  Node *head;
  Node *tail;

  /**
   * 队头插入元素
   */
  void addFromHead(int value) {
    Node *cur = new Node(value);
    if (head == nullptr) {
      head = cur, tail = cur;
    } else {
      cur->next = head;
      head->last = cur;

      head = cur;
    }
  }

  /**
   * 队尾插入元素
   */
  void addFromBottom(int value) {
    Node *cur = new Node(value);
    if (head == nullptr) {
      head = cur, tail = cur;
    } else {
      cur->last = tail;
      tail->next = cur;

      tail = cur;
    }
  }

  /**
   * 队头出队
   */
  int popFromHead() {
    if (head == nullptr) return -1;

    Node *cur = head;
    if (head == tail) {
      head = nullptr, tail = nullptr;
    } else {
      head = head->next;
      head->last = nullptr;
      cur->next = nullptr;
    }
    return cur->value;
  }

  /**
   * 队尾出队
   */
  int popFromBottom() {
    if (head == nullptr) return -1;

    Node *cur = head;
    if (head == tail) {
      head = nullptr, tail = nullptr;
    } else {
      tail = cur->last;
      tail->next = nullptr;
      cur->last = nullptr;
    }
    return cur->value;
  }

  int empty() { return head == nullptr; }
};

struct MyStack {
  DoubleEndsQueue *queue;
  MyStack() { queue = new DoubleEndsQueue(); }

  void push(int value) { queue->addFromHead(value); }

  int pop() { return queue->popFromHead(); }
  bool empty() { return queue->empty(); }
};
struct MyQueue {
  DoubleEndsQueue *queue;
  MyQueue() { queue = new DoubleEndsQueue(); }
  void push(int value) { queue->addFromBottom(value); }
  int poll() { return queue->popFromHead(); }
  bool empty() { return queue->empty(); }
};

class Solution2 {
 public:
  bool isEqual(int o1, int o2) {
    if (o1 == NULL && o2 != NULL) {
      return false;
    }
    if (o1 != NULL && o2 == NULL) {
      return false;
    }
    if (o1 == NULL && o2 == NULL) {
      return true;
    }
    return o1 == o2;
  }
};

int main() {
  Solution2 solution;
  int oneTestDataNum = 100;
  int value = 10000;
  int testTimes = 100000;
  for (int i = 0; i < testTimes; i++) {
    MyStack myStack;
    MyQueue myQueue;
    stack<int> stack;
    queue<int> queue;
    for (int j = 0; j < oneTestDataNum; j++) {
      int nums = (int)(rand() * value);
      if (stack.empty()) {
        myStack.push(nums);
        stack.push(nums);
      } else {
        if (rand() < 0.5) {
          myStack.push(nums);
          stack.push(nums);
        } else {
          int val = stack.top();
          stack.pop();
          int val2 = myStack.pop();
          if (!solution.isEqual(val2, val)) {
            cout << "Oops!" << endl;
          }
        }
      }

      int numq = (int)(rand() * value);
      if (queue.empty()) {
        myQueue.push(numq);
        queue.push(numq);
      } else {
        if (rand() < 0.5) {
          myQueue.push(numq);
          queue.push(numq);
        } else {
          int val = queue.front();
          queue.pop();
          int val2 = myQueue.poll();
          if (!solution.isEqual(val2, val)) {
            cout << "Oops!" << endl;
          }
        }
      }
    }
  }
  cout << "Finish!" << endl;
  return 0;
}